core/controller/src/main/scala/whisk/core/entitlement/KindRestrictor.scala

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.entitlement
+
+import whisk.common.Logging
+import whisk.common.TransactionId
+import whisk.core.ConfigKeys
+import whisk.core.entity.{Exec, Identity}
+import pureconfig.loadConfigOrThrow
+
+/**
+ * Restrict allowed kinds via provided whitelists
+ *
+ * @param whitelist set of default allowed kinds when not explicitly available via namespace limits
+ */
+protected[core] case class DefaultAllowedKinds(whitelist: Option[Set[String]] = None)
+
+/**
+ * A class for determining whether a given namespace is allowed to contain a given kind.
+ *
+ * @param allowedKinds a default set of allowed kinds
+ */
+class KindRestrictor(allowedKinds: DefaultAllowedKinds = loadConfigOrThrow[DefaultAllowedKinds](ConfigKeys.runtimes))(
+  implicit logging: Logging) {
+
+  logging.info(this, s"allowedKinds = $allowedKinds")(TransactionId.controller)
+
+  private def execKindAllowed(exec: Option[Exec], kinds: Set[String]): Boolean = {
+    exec match {
+      case Some(e) if kinds.contains(e.kind) => true
+      case _                                 => false
+    }
```

Comment:
This is equivalent to:

```scala
exec.map(e => kinds.contains(e.kind)).getOrElse(false)
```

In prose it says: If exec has a value, see if it is contained in the set and `get` that value. If exec has no value (`orElse`) default to false.